src/AutoRest.CSharp/Common/Output/Models/Types/HelperTypeProviders/ChangeTrackingListProvider.cs

```diff
@@ -0,0 +1,282 @@
+using AutoRest.CSharp.Generation.Writers;
+using AutoRest.CSharp.Input.Source;
+using AutoRest.CSharp.Output.Models;
+using AutoRest.CSharp.Output.Models.Shared;
+using AutoRest.CSharp.Output.Models.Types;
+
+namespace AutoRest.CSharp.Common.Output.Models.Types
+{
+    internal class ChangeTrackingListProvider : ExpressionTypeProvider
+    {
+        private static readonly Lazy<ChangeTrackingListProvider> _instance = new(() => new ChangeTrackingListProvider(Configuration.Namespace, null));
+        private readonly MethodSignature _ensureListSignature = new MethodSignature("EnsureList", null, null, MethodSignatureModifiers.Public, typeof(IList<>), null, Array.Empty<Parameter>());
+        private readonly MethodSignature _getEnumeratorSignature = new MethodSignature("GetEnumerator", null, null, MethodSignatureModifiers.Public, typeof(IEnumerator<>), null, Array.Empty<Parameter>());
+        private readonly TypeProvider _t;
+        private readonly FieldDeclaration _innerlListField;
+        private readonly TypeProvider _tArray;
+        private readonly Parameter _tParam;
+        private readonly Parameter _indexParam = new Parameter("index", null, typeof(int), null, ValidationType.None, null);
+        private VariableReference _innerList;
+
+        public static ChangeTrackingListProvider Instance => _instance.Value;
+
+        private ChangeTrackingListProvider(string defaultNamespace, SourceInputModel? sourceInputModel)
+            : base(defaultNamespace, sourceInputModel)
+        {
+            _innerlListField = new FieldDeclaration(FieldModifiers.Private, typeof(IList<>), "_innerList");
+            _innerList = new VariableReference(new CSharpType(typeof(IList<>)), _innerlListField.Declaration);
+            _t = new GenericParameterTypeProvider("T", DefaultNamespace, null);
+            _tArray = new GenericParameterTypeProvider("T[]", DefaultNamespace, null);
+            _tParam = new Parameter("item", null, _t.Type, null, ValidationType.None, null);
+            DeclarationModifiers = ClassSignatureModifiers.Internal;
+        }
+
+        protected override string DefaultName => "ChangeTrackingList";
+
+        protected override IEnumerable<Method> BuildConstructors()
+        {
+            yield return new Method(new ConstructorSignature(Type, null, null, MethodSignatureModifiers.Public, Array.Empty<Parameter>()), Snippets.EmptyStatement);
+            var iListParam = new Parameter("innerList", null, typeof(IList<>), null, ValidationType.None, null);
+            var iListSignature = new ConstructorSignature(Type, null, null, MethodSignatureModifiers.Public, new Parameter[] { iListParam });
+            var iListVariable = new ParameterReference(iListParam);
+            var iListbody = new MethodBodyStatement[]
+            {
+                new IfStatement(Snippets.NotEqual(iListVariable, Snippets.Null))
+                {
+                    new AssignValueStatement(_innerList, iListVariable)
```

Comment:
There is a method that we could do this in a more semantic way
```suggestion
                    Assign(_innerList, iListVariable)
```
you need to using static Snippets; <- VS should have intellisence to automatically complete this because I do not remember the full namespace here

Reply:
The generic function does not like this for some reason.  New AssignValueStatement seems to work fine though.

Reply:
Well just because their type is different and we should do this:
```suggestion
                    Assign<ValueExpression>(_innerList, iListVariable)
```
or we could change the type of field `_innerList` to `ValueExpression` because usually we do not need an explicit `VariableReference`.